Add Hover Active State

joshmedeski opened the pull request:
For https://github.com/Medology/stdcheck.com/issues/1990

- Added a makefile for spacemacs (quick commands)
- Added `npm`, `bower` and `gulp` docker scripts
- Added styling for the hover/active button states

__Preview:__
![ffab-hover-active](https://cloud.githubusercontent.com/assets/2036594/21686089/b5b4089e-d329-11e6-8bbe-3679e13b0403.gif)

Commit:
:wrench: Add makefile and docker tooling

Commit:
✨ Add hover-active states to the button

joshmedeski commented:
@atroppe can you take a look at this for me?

atroppe commented on index.html:
```diff
@@ -64,8 +64,11 @@
       <br>
-      <a href="#" class="[ button ][ ffab static fa-arrow-right ]">No Transition (Static)</a>
+      <a href="#" class="[ button ][ ffab fa-arrow-right ]">With Hover/Active</a>
+      <a href="#" class="[ button ][ ffab fa-arrow-right no-hover-active ]">Without Hover/Active</a>
```
If there is a way to turn on/off the hover effect and the active state individually, I think it would be useful.
Also it seems like there might be a way to use the sass setting to determine whether or not to include it, without have to add a class.  Is that possible?

joshmedeski replied:
That's what the code is already doing. If you set `$ffab-has-hover-active-state` to `false` this feature is turned off. The `no-hover-active` is a way to disable the feature on an individual component.

I haven't thought about adding the ability to add the feature on an individual component. Do you think that would be valuable functionality?

atroppe replied:
The short answer, yes I would definitely want to be able to add one or the other, though not necessarily on the individual level.  The long answer, see below...

atroppe commented on src/_ffab-hover-active.scss:
```diff
@@ -0,0 +1,24 @@
+
+@mixin ffab-hover-active {
+  @if $ffab-has-hover-active-state {
+    .button:not(.disabled):not(.no-hover-active) {
```
(following my previous comment) wouldn't you be able to exclude the :not(.no-hover-active), because this is already in an if statement?

joshmedeski replied:
This is to manually disable the active-hover feature on an individual component. It feels valuable if there is an isolated situation were you don't want to include the feature. I'd like to see your response to my first comment to decide how we want to update this code.

atroppe commented on src/_ffab-hover-active.scss:
```diff
@@ -0,0 +1,24 @@
+$ffab-has-hover-active-state: true !default;
+$ffab-hover-active-transition-speed: 150ms !default;
+$ffab-hover-value: -1px !default;
+$ffab-active-value: 1px !default;
```
It seems like it is shifting more than 1px.  any idea whats going on?

joshmedeski replied:
It's starting 1px up so it's going 2px down (1px to get back to it's original position and one more pixel with the transform).

atroppe replied:
I would rather shift it only on active, since there is already a distinct hover effect that is well done, and more "business" detracts from that.
This might be a good use case for separating it into different options:
1. `ffab`
The original hover effect only. Possibly adding an active state where the icon shifts to the right for `ffab` and to the left for `ffab before`.  Maybe add the active state only with the combinator `active` ?
2. `ffab push`
The new active state where it shifts down on active, and up on hover, without the original hover effect.
3. `ffab touchscreen`
For apps whose primary purpose is mobile/tablet use, where the button jumps towards you in the "material design" fashion.

My names aren't so great but you get the idea.

atroppe reviewed (commented):
I can't help with the scripts, but I left a few comments. (although it's already pretty awesome)

Commit:
💪 Make all links void and point before arrows left

Commit:
✨ Move ffab icon on :active state

Removing the hover/active idea and just moving the ffab icon on the :active state

Commit:
:muscle: Alphabetize imports and includes

joshmedeski commented:
Thanks for the feedback @atroppe. I've rewritten the animation to do two things.

- NOT do anything on hover (let the wipe animation be the focus)
- Animate the icon to the side on `:active`

This is simple and doesn't overstretch the role of this project. A simple addition to add another layer of effect that can be easily turned on/off if needed.

__Preview:__
![ffab-hover](https://cloud.githubusercontent.com/assets/2036594/21787822/f7c86422-d690-11e6-9dba-9eadb2231f27.gif)

atroppe commented on src/_ffab-active.scss:
```diff
@@ -0,0 +1,25 @@
+
+@mixin ffab-active {
+  @if $ffab-has-active-state {
+    .ffab:not(.disabled):not(.no-hover-active) {
```
since you have separated the hover and active effects, do you still want to use this class name?  or does it disable the active state as well as the hover effect?

joshmedeski replied:
Good catch, I need to update this as well.

atroppe reviewed (commented):
one more question

atroppe commented:
Looks great!

Commit:
:muscle: Adjust class that disables active state
